getmntinfo() supports, according to mount.h on FreeBSD current:
440 /*
441  * Flags for various system call interfaces.
442  *
443  * waitfor flags to vfs_sync() and getfsstat()
444  */
445 #define MNT_WAIT        1       /* synchronously wait for I/O to complete */
446 #define MNT_NOWAIT      2       /* start all I/O, but do not wait for it */
447 #define MNT_LAZY        3       /* push data not written by filesystem syncer */
448 #define MNT_SUSPEND     4       /* Suspend file system after sync */

flags=0 is not supported.

--- src/corelib/io/qstorageinfo_unix.cpp.orig	2017-05-29 21:24:28.134722000 +0200
+++ src/corelib/io/qstorageinfo_unix.cpp	2017-05-29 21:25:08.844553000 +0200
@@ -194,7 +194,7 @@
 #if defined(Q_OS_BSD4)
 
 inline QStorageIterator::QStorageIterator()
-    : entryCount(::getmntinfo(&stat_buf, 0)),
+    : entryCount(::getmntinfo(&stat_buf, MNT_NOWAIT)),
       currentIndex(-1)
 {
 }
